pkg/generate/appjson/appjson.go

```diff
@@ -0,0 +1,388 @@
+	Website     string
+	Logo        string
+	SuccessURL  string `json:"success_url"`
+	Scripts     map[string]string
+	Env         map[string]EnvVarOrString
+	Formation   map[string]Formation
+	Image       string
+	Addons      []string
+	Buildpacks  []Buildpack
+}
+
+type Generator struct {
+	LocalPath string
+	Name      string
+}
+
+// Generate accepts a path to an app.json file and generates a template from it
+func (g *Generator) Generate(body []byte) (*templateapi.Template, error) {
+	appJSON := &AppJSON{}
+	if err := json.Unmarshal(body, appJSON); err != nil {
+		return nil, err
+	}
+
+	glog.V(4).Infof("app.json: %#v", appJSON)
+
+	warnings := make(map[string][]string)
+	warnUnusableAppJSONElements("app.json", appJSON, warnings)
+
+	name := g.Name
+	if len(name) == 0 && len(g.LocalPath) > 0 {
+		name = filepath.Base(g.LocalPath)
+	}
+
+	template := &templateapi.Template{}
+	template.Name = name
+	template.Annotations = make(map[string]string)
+	template.Annotations["openshift.io/website"] = appJSON.Website
+	template.Annotations["k8s.io/display-name"] = appJSON.Name
+	template.Annotations["k8s.io/description"] = appJSON.Description
+	template.Annotations["tags"] = strings.Join(appJSON.Keywords, ",")
+	template.Annotations["iconURL"] = appJSON.Logo
+
+	// create parameters and environment for containers
+	allEnv := make(app.Environment)
+	for k, v := range appJSON.Env {
+		if v.EnvVar != nil {
+			allEnv[k] = fmt.Sprintf("${%s}", k)
+		}
+	}
+	envVars := allEnv.List()
+	for _, v := range envVars {
+		env := appJSON.Env[v.Name]
+		if env.EnvVar == nil {
+			continue
+		}
+		e := env.EnvVar
+		displayName := v.Name
+		displayName = strings.Join(strings.Split(strings.ToLower(displayName), "_"), " ")
+		displayName = strings.ToUpper(displayName[:1]) + displayName[1:]
+		param := templateapi.Parameter{
+			Name:        v.Name,
+			DisplayName: displayName,
+			Description: e.Description,
+			Value:       e.Value,
+		}
+		switch e.Generator {
+		case "secret":
+			param.Generate = "expression"
+			param.From = "[a-zA-Z0-9]{14}"
+		}
+		if len(param.Value) == 0 && e.Default != nil {
+			switch t := e.Default.(type) {
+			case string:
+				param.Value = t
+			case float64, float32:
+				out, _ := json.Marshal(t)
+				param.Value = string(out)
+			}
+		}
+		template.Parameters = append(template.Parameters, param)
+	}
+
+	if len(appJSON.Formation) == 0 {
+		glog.V(4).Infof("No formation in app.json, adding a default web")
+		// TODO: read Procfile for command?
+		appJSON.Formation = map[string]Formation{
+			"web": {
+				Quantity: 1,
+			},
+		}
+	}
+
+	formations := sets.NewString()
+	for k := range appJSON.Formation {
+		formations.Insert(k)
+	}
+
+	imageGen := app.NewImageRefGenerator()
+
+	buildPath := appJSON.Repository
+	if len(buildPath) == 0 && len(g.LocalPath) > 0 {
+		buildPath = g.LocalPath
+	}
+	if len(buildPath) == 0 {
+		return nil, fmt.Errorf("app.json did not contain a repository URL and no local path was specified")
+	}
```

Comment:
so what actually happens if the content is from the local filesystem, not a repo?